inspirehep/modules/workflows/views.py

```diff
@@ -376,3 +400,184 @@ def robotupload_callback():
         responses[recid] = response
 
     return jsonify(responses)
+
+
+def error_handler(f):
+    """Decorator to handle callback exceptions."""
+
+    @wraps(f)
+    def inner(*args, **kwargs):
+        try:
+            return f(*args, **kwargs)
+        except CallbackMalformedError as error:
+            return make_response(
+                dict(error_code=error.error_code,
+                     message=error.description),
+                code=error.code
+            )
+        except CallbackWorkflowNotFoundError as error:
+            return make_response(
+                dict(error_code=error.error_code,
+                     message=error.description),
+                code=error.code
+            )
+        except CallbackValidationError as error:
+            return make_response(
+                dict(error_code=error.error_code,
+                     message=error.description,
+                     workflow=error.workflow),
+                code=error.code
+            )
+    return inner
+
+
+def validate_workflow_payload(keys):
+    """Validate and get the workflow data from the request.
+
+    This decorator will first check if all the given keys are part of
+    the request's payload, after will validate the ``metadata`` against
+    the ``hep`` JSONSchema and finally it will pass the ``workflow_data``
+    kwarg.
+
+    Args:
+        keys (list(str)): the keys to check if they exist.
+
+    Raises:
+        CallbackMalformedError: if any of the ``keys`` is missing from
+            the request payload.
+        CallbackValidationError: if the workflow ``metadata`` is not valid
+            against ``hep`` JSONSchema.
+    """
+    def validate_wrap(f):
+        """Decorator to handle exceptions."""
+        @wraps(f)
+        def inner(*args, **kwargs):
+            workflow_data = request.get_json()
+
+            # Check for missing keys
+            for key in keys:
+                if workflow_data.get(key) is None:
+                    raise CallbackMalformedError(key)
+
+            # Check for validation errors
+            try:
+                validate(schema='hep', data=workflow_data['metadata'])
+            except ValidationError:
+                workflow_data['_extra_data']['validation_errors'] = \
+                    validation_exception_to_validation_errors(
+                        workflow_data['metadata'], 'hep')
+                workflow_data['_extra_data']['callback_url'] = \
+                    resolve_validation_callback_url()
+                raise CallbackValidationError(workflow_data)
+
+            return f(workflow_data=workflow_data, *args, **kwargs)
+        return inner
+    return validate_wrap
+
+
+class ResolveValidationResource(MethodView):
+    """Resolve validation error callback."""
+
+    @error_handler
+    @validate_workflow_payload(['metadata', '_extra_data', 'id'])
+    def put(self, workflow_data, **kwargs):
```

Comment:
Can you avoid using a decorator that changes the signature?

Reply:
Also, is the kwargs needed?

Reply:
the decorator is to help us deal with other workflow callback cases in the future (ala ``@pass_record``) indeed ``kawrgs`` is not needed

Reply:
Can we use a function instead of a decorator?

Reply:
IMO is cleaner to have a decorator in such cases, you can easily change the validation for the workflow payload and you have a workflow object ready. This is how ``records-rest`` is working, it'll be a good start to keep everything tidy and reusable for the workflows  REST.